core/block/object/objectcreator/creator.go

```diff
@@ -80,109 +69,40 @@ func (s *service) Init(a *app.App) (err error) {
-) (id string, newDetails *types.Struct, err error) {
-	if createState == nil {
-		createState = state.NewDoc("", nil).(*state.State)
-	}
-	startTime := time.Now()
-	// priority:
-	// 1. details
-	// 2. createState
-	// 3. createState details
-	// 4. default object type by smartblock type
-	if len(objectTypeKeys) == 0 {
-		objectTypeKeys = []domain.TypeKey{bundle.TypeKeyPage}
-	}
-	sbType := objectTypeKeysToSmartBlockType(objectTypeKeys)
-
-	createState.SetDetailAndBundledRelation(bundle.RelationKeySpaceId, pbtypes.String(spc.Id()))
-
-	ev := &metrics.CreateObjectEvent{
-		SetDetailsMs: time.Since(startTime).Milliseconds(),
-	}
-
-	ctx = context.WithValue(ctx, eventCreate, ev)
-	initFunc := func(id string) *smartblock.InitContext {
-		createState.SetRootId(id)
-		return &smartblock.InitContext{
-			Ctx:            ctx,
-			ObjectTypeKeys: objectTypeKeys,
-			State:          createState,
-			RelationKeys:   generateRelationKeysFromState(createState),
-			SpaceID:        spc.Id(),
-		}
-	}
-
-	sb, err := createSmartBlock(ctx, spc, initFunc, createState, sbType)
-	if err != nil {
-		return "", nil, err
-	}
-
-	newDetails = sb.CombinedDetails()
-	id = sb.Id()
-
-	if sbType == coresb.SmartBlockTypeObjectType {
-		objecttype.UpdateLastUsedDate(spc, s.objectStore, []domain.TypeKey{
-			domain.TypeKey(strings.TrimPrefix(pbtypes.GetString(newDetails, bundle.RelationKeyUniqueKey.String()), addr.ObjectTypeKeyToIdPrefix)),
-		})
-	} else if pbtypes.GetInt64(newDetails, bundle.RelationKeyOrigin.String()) == int64(model.ObjectOrigin_none) {
-		objecttype.UpdateLastUsedDate(spc, s.objectStore, objectTypeKeys)
-	}
-
-	ev.SmartblockCreateMs = time.Since(startTime).Milliseconds() - ev.SetDetailsMs - ev.WorkspaceCreateMs - ev.GetWorkspaceBlockWaitMs
-	ev.SmartblockType = int(sbType)
-	ev.ObjectId = id
-	metrics.SharedClient.RecordEvent(*ev)
-	return id, newDetails, nil
-}
-
-func getUniqueKeyOrGenerate(sbType coresb.SmartBlockType, details *types.Struct) (domain.UniqueKey, error) {
-	uniqueKey := pbtypes.GetString(details, bundle.RelationKeyUniqueKey.String())
-	if uniqueKey == "" {
-		return domain.NewUniqueKey(sbType, bson.NewObjectId().Hex())
-	}
-	return domain.UnmarshalUniqueKey(uniqueKey)
-}
-
 // TODO Add validate method
 type CreateObjectRequest struct {
 	Details       *types.Struct
 	InternalFlags []*model.InternalFlag
 	TemplateId    string
 	ObjectTypeKey domain.TypeKey
+	UniqueKey     string
 }
 
+// CreateObject is high-level method for creating new objects
 func (s *service) CreateObject(ctx context.Context, spaceID string, req CreateObjectRequest) (id string, details *types.Struct, err error) {
 	space, err := s.spaceService.Get(ctx, spaceID)
 	if err != nil {
 		return "", nil, fmt.Errorf("get space: %w", err)
 	}
-	return s.CreateObjectInSpace(ctx, space, req)
+
+	if req.ObjectTypeKey.String() == "" {
+		objectTypeKey, err := domain.GetTypeKeyFromRawUniqueKey(req.UniqueKey)
```

Comment:
what's the purpose of this? 

CreateObjectUsingObjectUniqueTypeKey was pretty specific method

Reply:
We are including specific fields of CreateObject request in CreateObjectRequest structure. We can make specific method for each field: CreateObjectUsingTemplate or CreateObjectWithDetails. Maybe we should leave it more general and use single CreateObject method in all places?